src/scandeval/generation.py

```diff
@@ -165,79 +192,215 @@ def generate_single_iteration(
-    )
-    all_preds: list[str | list[str]] = list()
+        generation_config = GenerationConfig(
+            max_new_tokens=dataset_config.max_generated_tokens,
+            do_sample=False,
+            stopping_criteria=stopping_criteria,
+            output_scores=True,
+            return_dict_in_generate=True,
+            bos_token_id=tokenizer.bos_token_id,
+            eos_token_id=tokenizer.eos_token_id,
+            pad_token_id=tokenizer.pad_token_id,
+        )
 
-    batch_size = 1 if isinstance(model, OpenAIModel) else benchmark_config.batch_size
-    dataloader = DataLoader(
-        dataset=torch_dataset,
-        batch_size=batch_size,
-        shuffle=False,
-        num_workers=4,
-        collate_fn=data_collator,
-    )
+        # Sort the non_cached dataset by the length of the text, to minimise the amount
+        # of padding that needs to be added, speeding up generation
+        non_cached_dataset = non_cached_dataset.add_column(
+            name="length", column=[len(x) for x in non_cached_dataset["text"]]
+        )
+        non_cached_dataset = non_cached_dataset.sort("length", reverse=False)
+
+        # Enable batching by building a dataloader. The dataloader cannot deal with
+        # text columns, so we create a copy of the dataset without these
+        torch_dataset = non_cached_dataset.with_format("torch").remove_columns(
+            [
+                column
+                for column in non_cached_dataset.column_names
+                if column != "input_ids"
+            ]
+        )
+
+        batch_size = (
+            1 if isinstance(model, OpenAIModel) else benchmark_config.batch_size
+        )
+        dataloader = DataLoader(
+            dataset=torch_dataset,
+            batch_size=batch_size,
+            shuffle=False,
+            num_workers=4,
+            collate_fn=data_collator,
+        )
 
-    for batch_idx, batch in enumerate(tqdm(dataloader, leave=False)):
-        # Generate the completions of the documents in the batch
-        with warnings.catch_warnings():
-            warnings.simplefilter("ignore", category=UserWarning)
-            inputs = batch["input_ids"].to(model.device)
-            with torch.inference_mode():
-                model_output: ModelOutput = model.generate(
-                    inputs=inputs, generation_config=generation_config
+        # Generate the completions for the non-cached examples
+        for batch_idx, batch in enumerate(tqdm(dataloader, leave=False)):
+            # Generate the completions of the documents in the batch
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore", category=UserWarning)
+                inputs = batch["input_ids"].to(model.device)
+                with torch.inference_mode():
+                    model_output: ModelOutput = model.generate(
+                        inputs=inputs, generation_config=generation_config
+                    )
```

Comment:
I believe you can do:
```
            with (
              warnings.catch_warnings(),
              torch.inference_mode()
              ):
              ...
```
Due to:
https://docs.python.org/3.10/whatsnew/3.10.html#parenthesized-context-managers